File: BOJ(Java 공부용)/15686.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

import static java.lang.Integer.min;
import static java.lang.Math.abs;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int ans = (int) 1e9;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        int[][] board = new int[n][n];
        ArrayList<p> v = new ArrayList<>();
        for(int i=0;i<n;i++) {
            st = new StringTokenizer(br.readLine());
            for(int j=0;j<n;j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
                if(board[i][j] == 2) v.add(new p(i,j));
            }
        }

        boolean[] visited = new boolean[v.size()];
        go(0,0,m,v,visited,board,n);
        bw.write(ans+"\n");
        bw.flush();
        bw.close();
    }
    private static void go(int here,int cnt,int depth,ArrayList<p> v, boolean[] visited,int[][] board, int n) {
        if(cnt == depth) {
            bfs(v,visited,board,n);
            return;
        }
        for(int i=here;i<v.size();i++) {
            if(!visited[i]) {
                visited[i] = true;
                go(i+1,cnt+1,depth,v,visited,board,n);
                visited[i] = false;
            }
        }
    }
    private static void bfs(ArrayList<p> v,boolean[] visited,int[][] board, int n) {
        ArrayList<p> temp = new ArrayList<>();
        for(int i=0;i<v.size();i++) {
            if(visited[i]) temp.add(new p(v.get(i).x,v.get(i).y));
        }
        ArrayList<p> home = new ArrayList<>();
        for(int i=0;i<n;i++) {
            for(int j=0;j<n;j++) {
                if(board[i][j] == 1) {
                    home.add(new p(i,j));
                }
            }
        }
        int tempans = 0;
        for(int i=0;i<home.size();i++) {
            int tempdist = (int) 1e9;
            for(int j=0;j<temp.size();j++) {
                tempdist = min(tempdist,abs(home.get(i).x-temp.get(j).x)+abs(home.get(i).y-temp.get(j).y));
            }
            tempans += tempdist;
        }
        ans = min(ans,tempans);
    }
    private static class p {
        int x,y;
        private p(int x,int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: BOJ(Python 공부용)/15686.py
import sys

input = sys.stdin
sys.setrecursionlimit(500000)

n, m = map(int, input.readline().split())
board = [[int(x) for x in input.readline().split()] for _ in range(n)]
v = []
ans = int(1e9)


def bfs():
    global ans
    temp = []
    for i in range(len(v)):
        if visited[i]:
            temp.append((v[i][0], v[i][1]))

    home = []
    for i in range(n):
        for j in range(n):
            if board[i][j] == 1:
                home.append((i, j))

    tempans = 0
    for i in range(len(home)):
        tempdist = int(1e9)
        for j in range(len(temp)):
            tempdist = min(tempdist, abs(home[i][0] - temp[j][0]) + abs(home[i][1] - temp[j][1]))
        tempans += tempdist

    ans = min(ans, tempans)


def dfs(here, cnt, depth):
    if cnt == depth:
        bfs()
        return
    for i in range(here, len(v), 1):
        if not visited[i]:
            visited[i] = True
            dfs(i + 1, cnt + 1, depth)
            visited[i] = False


for i in range(n):
    for j in range(n):
        if board[i][j] == 2:
            v.append((i, j))

visited = [False] * len(v)
dfs(0, 0, m)
print(ans)